### PR TITLE
feat: adjustable rule_group interval

### LINKED ISSUE #367: Add ability to specify rule group interval
I'm wanting to override the global default interval value with a [rule_group](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#rule_group) interval value. When generating the rules via sloth I do not see the ability to set this.

### PR DESCRIPTION
**Pull request by wbollock**
This allows for more expensive SLO recording rules by letting users create a custom Prometheus `rule_group.interval`.

https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#rule_group

Now instead of all rule groups using the default global evaluation interval, a custom interval can be set on all sets of recording rules for an SLO.

If no `interval` is set then the global default will be assumed, matching current Sloth behavior.

Resolves: https://github.com/slok/sloth/issues/367

### CODE REVIEW

File: internal/prometheus/model.go [hunk 1]
```diff
@@ -39,16 +39,17 @@ type AlertMeta struct {
...
-	TimeWindow      time.Duration     `validate:"required"`
-	Objective       float64           `validate:"gt=0,lte=100"`
-	Labels          map[string]string `validate:"dive,keys,prom_label_key,endkeys,required,prom_label_value"`
-	PageAlertMeta   AlertMeta
-	TicketAlertMeta AlertMeta
+	ID                string `validate:"required,name"`
+	Name              string `validate:"required,name"`
+	Description       string
+	Service           string `validate:"required,name"`
+	RuleGroupInterval string
```
Comment (wbh1) [inline internal/prometheus/model.go]: Why not make this `time.Duration`, too?

**Review by wbh1 — COMMENTED** [review]
Overall, looks good! Since Sloth produces 3 rules groups, I think it'd be nice to support specifying separate intervals for each of them, though. (`sloth-slo-meta-recordings`, `sloth-slo-alerts`, and `sloth-slo-sli-recordings` are the 3 prefixes). For example, you may want an interval of `2m` on the SLI recording rules but still `1m` on the `alerts`.

File: internal/prometheus/model.go (hunk 1 above)
Comment (wbollock) [reply internal/prometheus/model.go]: Originally I did but I had to convert it to a string to parse the duration anyway into the `prommodel` format. I can switch back!

**Comment by wbollock** [comment]
@wbh1 commits added for per rule group intervals:
a94fee994c770e741571f9d70efeee51890cd5f1
ca78450231e2ded7baa2f995b3ff7f97af933be0
dc3994be548f1598c633cc19de25b8b45a3dc416

File: internal/prometheus/storage.go [hunk 2]
```diff
@@ -49,24 +49,95 @@ func (i IOWriterGroupedRulesYAMLRepo) StoreSLOs(ctx context.Context, slos []Stor
...
-			ruleGroups.Groups = append(ruleGroups.Groups, ruleGroupYAMLv2{
-				Name:  fmt.Sprintf("sloth-slo-meta-recordings-%s", slo.SLO.ID),
-				Rules: slo.Rules.MetadataRecRules,
-			})
+			// if either of these aren't empty we'll be adding a custom rule interval
+			if slo.SLO.RuleGroupInterval.String() != "0s" || slo.SLO.MetadataRulesInterval.String() != "0s" {
+				var ruleGroupIntervalDuration prommodel.Duration
+				var err error
+
+				// if we have a valid meta rule rule_group interval, use that firs
```
Comment (wbh1) [inline internal/prometheus/storage.go]:
```suggestion
				// if we have a valid meta rule rule_group interval, use that first and overwrite any generic ones
```

File: internal/prometheus/storage.go [hunk 3]
```diff
@@ -49,24 +49,95 @@ func (i IOWriterGroupedRulesYAMLRepo) StoreSLOs(ctx context.Context, slos []Stor
...
 		if len(slo.Rules.AlertRules) > 0 {
-			ruleGroups.Groups = append(ruleGroups.Groups, ruleGroupYAMLv2{
-				Name:  fmt.Sprintf("sloth-slo-alerts-%s", slo.SLO.ID),
-				Rules: slo.Rules.AlertRules,
-			})
+			if slo.SLO.RuleGroupInterval.String() != "0s" || slo.SLO.AlertRulesInterval.String() != "0s" {
+				var ruleGroupIntervalDuration prommodel.Duration
+				var err error
+
+				// if we have a valid meta rule rule_group interval, use that firs
```
Comment (wbh1) [inline internal/prometheus/storage.go]:
```suggestion
				// if we have a valid meta rule rule_group interval, use that first and overwrite any generic ones
```

File: internal/prometheus/storage.go [hunk 4]
```diff
@@ -49,24 +49,95 @@ func (i IOWriterGroupedRulesYAMLRepo) StoreSLOs(ctx context.Context, slos []Stor
...
+					RuleGroupInterval: ruleGroupIntervalDuration,
+					Rules:             slo.Rules.SLIErrorRecRules,
+				})
+			} else {
+				ruleGroups.Groups = append(ruleGroups.Groups, ruleGroupYAMLv2{
+					Name:  fmt.Sprintf("sloth-slo-sli-recordings-%s", slo.SLO.ID),
+					Rules: slo.Rules.SLIErrorRecRules,
+				})
+
+			}
```
Comment (wbh1) [inline internal/prometheus/storage.go]: I think you can simplify this with a `switch` statement like:

```go
    group := ruleGroupYAMLv2{
					Name:             fmt.Sprintf("sloth-slo-sli-recordings-%s", slo.SLO.ID),
					Rules:             slo.Rules.SLIErrorRecRules,
				}
	switch {
	case AlertRulesInterval.String() != "0s":
		ruleGroupIntervalDuration, err = time.ParseDuration(AlertRulesInterval.String())
        if err != nil {
            return fmt.Errorf("could not parse rule_group interval duration for alerts %w", err)
        } else {
            group.RuleGroupInterval = ruleGroupIntervalDuration
        }
	case RuleGroupInterval.String() != "0s":
		ruleGroupIntervalDuration, err = time.ParseDuration(RuleGroupInterval.String())
        if err != nil {
            return fmt.Errorf("could not parse default ("all") rule_group interval duration %w", err)
        } else {
            group.RuleGroupInterval = ruleGroupIntervalDuration
        }
	}

    ruleGroups.Groups = append(ruleGroups.Groups, group)
```

**Review by wbh1 — APPROVED** [review]
Approving as it looks good as-is, but left some comments on form.

**Pull request merged**